### PR TITLE
[dy] Update oauth sign on and fix oauth sign on with REQUIRE_USER_PERMISSIONS

### PR DESCRIPTION
**Pull request by dy46**
# Description

Signing in with an Oauth provider wasn't working properly when `REQUIRE_USER_PERMISSIONS` was enabled, so fixing that in this PR. Also, I added some refactoring changes to the active directory oauth provider, and to the oauth resource to clean some things up.

# How Has This Been Tested?

- [x] Tested locally with active directory
- [x] Tested locally with active directory with REQUIRE_USER_PERMISSIONS=1
- [x] Unit tests

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] If new documentation has been added, relative paths have been added to the appropriate section of `docs/mint.json`

cc: @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/authentication/providers/ghe.py [hunk 1]
```diff
@@ -39,30 +39,29 @@ def __validate(self):
...
-            query_strings = []
-            for k, v in query.items():
-                query_strings.append(f'{k}={v}')
+        base_url = get_base_url(redirect_uri)
+        redirect_uri_query = dict(
+            provider=self.provider,
+            redirect_uri=redirect_uri,
+        )
+        query = dict(
+            client_id=os.getenv(GHE_CLIENT_ID_ENV_VAR),
```
Comment (wangxiaoyou1993) [inline mage_ai/authentication/providers/ghe.py:48]: this can be replaced with `self.client.id`

File: mage_ai/api/policies/OauthPolicy.py [hunk 2]
```diff
@@ -23,6 +72,7 @@ class OauthPolicy(BasePolicy):
 OauthPolicy.allow_actions(
     [
         constants.DETAIL,
+        constants.LIST,
```
Comment (wangxiaoyou1993) [inline mage_ai/api/policies/OauthPolicy.py:75]: Question: Is it safe to return the oauth urls publicly? Are there any sensitive data in the response?

File: mage_ai/authentication/providers/okta.py [hunk 3]
```diff
@@ -36,33 +36,32 @@ def __validate(self):
...
-                ),
-                response_mode='query',
-                response_type='code',
-                scope='openid email profile',
-                state=uuid.uuid4().hex,
-            )
-            query_strings = []
-            for k, v in query.items():
-                query_strings.append(f'{k}={v}')
+    def get_auth_url_response(self, redirect_uri: str = None, **kwargs) -> Dict:
```
Comment (wangxiaoyou1993) [inline mage_ai/authentication/providers/okta.py:39]: This is a comment for the `__init__` method in this class. You might need to call `__validate` before settings the self.hostname.

File: mage_ai/api/policies/OauthPolicy.py (hunk 2 above)
Comment (dy46) [reply mage_ai/api/policies/OauthPolicy.py:75]: no there shouldn't be any sensitive data in the initial oauth redirect url

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**